dpu_utils/utils/richpath.py

```diff
@@ -255,17 +260,34 @@ def read_as_binary(self) -> bytes:
 
     def __cache_file_locally(self, num_retries: int=1) -> LocalPath:
         cached_file_path = os.path.join(self.__cache_location, self.__container_name, self.path)
-        if not os.path.isfile(cached_file_path):
-            try:
-                os.makedirs(os.path.dirname(cached_file_path), exist_ok=True)
-                self.__blob_service.get_blob_to_path(self.__container_name, self.path, cached_file_path)
-            except Exception as e:
-                if os.path.exists(cached_file_path):
-                    os.remove(cached_file_path)   # On failure, remove the cached file, if it exits.
-                if num_retries == 0:
-                    raise e
-                else:
-                    self.__cache_file_locally(num_retries-1)
+        cached_file_path_etag = cached_file_path+'.etag'  # Create an .etag file containing the object etag
+        old_etag = None
+        if os.path.exists(cached_file_path_etag):
+            with open(cached_file_path_etag) as f:
+                old_etag = f.read()
+
+        try:
+            os.makedirs(os.path.dirname(cached_file_path), exist_ok=True)
+            # The next invocation to the blob service may fail and delete the current file. Store it elsewhere
+            new_filepath = cached_file_path+'.new'
+
+            blob = self.__blob_service.get_blob_to_path(self.__container_name, self.path, new_filepath,
+                                                        if_none_match=old_etag)
+            os.rename(new_filepath, cached_file_path)
+            with open(cached_file_path_etag, 'w') as f:
+                f.write(blob.properties.etag)
+        except AzureHttpError as aze:
+            os.remove(new_filepath)
+            if aze.error_code != 'ConditionNotMet':
+                raise aze

```

Comment:
This should just be `raise` (which means re-raise, instead of raising a new exception in the current place). See https://wiki.python.org/moin/HandlingExceptions for details on this.